### PR TITLE
[js/web/training] Added parameters methods

### PR DESCRIPTION
**Pull request by carzh**
### Description
* Implemented: `getParametersSize`, `getContiguousParameters` (equivalent to copyParametersToBuffer), and `loadParametersBuffer` (equivalent to copyParametersFromBuffer)
  * as part of these changes, getParametersSize was added to the TrainingSession interface so that users know what size buffer to create for loadParametersBuffer
  * The parameters methods in the interface were modified to take in a Float32Array instead

### Motivation and Context
* part of the work for implementing web bindings for training
* enables federated learning in the web
* previous  PR: #18006

### CODE REVIEW

File: js/common/lib/training-session-impl.ts [hunk 1]
```diff
@@ -176,12 +176,22 @@ export class TrainingSession implements TrainingSessionInterface {
...
+    if (array.length !== paramsSize) {
+      throw new Error(
+          'Size of the buffer passed into loadParametersBuffer must match the number of parameters in ' +
+          'the model. Please use getParametersSize method to check.');
+    }
+    return this.handler.loadParametersBuffer(array, trainableOnly);
+  }
+
+  async getContiguousParameters(trainableOnly: boolean): Promise<OnnxValue> {
+    return this.handler.getContiguousParameters(trainableOnly);
```
Comment (askhade) [inline js/common/lib/training-session-impl.ts:194]: user does not have an option to pass a preallocated buffer? Is this a limitation from CAPI or we decided on the js side?

File: js/common/lib/training-session.ts [hunk 2]
```diff
@@ -49,21 +50,29 @@ export interface TrainingSession {
...
    * Copies from a buffer containing parameters to the TrainingSession parameters.
    *
    * @param buffer - buffer containing parameters
    * @param trainableOnly - True if trainable parameters only to be modified, false otherwise.
    */
-  loadParametersBuffer(array: Uint8Array, trainableOnly: boolean): Promise<void>;
+  loadParametersBuffer(array: Float32Array, trainableOnly: boolean): Promise<void>;
 
   /**
    * Copies from the TrainingSession parameters to a buffer.
```
Comment (askhade) [inline js/common/lib/training-session.ts:70]: Copies all the parameters to a contiguous buffer.

File: js/common/lib/training-session.ts [hunk 3]
```diff
@@ -49,21 +50,29 @@ export interface TrainingSession {
...
    *
    * @param trainableOnly - True if trainable parameters only to be copied, false othrwise.
    * @returns A promise that resolves to a buffer of the requested parameters.
```
Comment (askhade) [inline js/common/lib/training-session.ts:73]: When set to true only trainable parameters (parameters for which requires_grad is set to true) are copied.

File: js/common/lib/training-session.ts [hunk 4]
```diff
@@ -49,21 +50,29 @@ export interface TrainingSession {
...
    */
-  getContiguousParameters(trainableOnly: boolean): Promise<Uint8Array>;
+  getContiguousParameters(trainableOnly: boolean): Promise<OnnxValue>;
```
Comment (askhade) [inline js/common/lib/training-session.ts:75]: what is the default value?

Comment (askhade) [reply js/common/lib/training-session.ts:75]: documentation should also mention default behavior

File: js/common/lib/training-session.ts [hunk 5]
```diff
@@ -49,21 +50,29 @@ export interface TrainingSession {
   // #endregion
 
   // #region copy parameters
+
+  /**
+   * Retrieves the size of all parameters for the training state.
```
Comment (askhade) [inline js/common/lib/training-session.ts:55]: You can use the description from the C API header file

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 6]
```diff
@@ -16,6 +16,22 @@ const NO_TRAIN_FUNCS_MSG =
     'functionality, and make sure that all the correct artifacts are built & moved to the correct folder if ' +
     'using a custom build. Check https://onnxruntime.ai/docs/build/web.html for more information.';
 
+/**
+ * Runs the checkLastError function which will throw an error, if the provided error code matches the specified
+ * pattern for an error code.
+ * @param errCode number to evaluated for if it's an erro
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:22]:
```suggestion
 * @param errCode number to evaluated for if it's an error
```

File: js/common/lib/training-session-impl.ts (hunk 1 above)
Comment (carzh) [reply js/common/lib/training-session-impl.ts:194]: The user would not be able to pre-allocate a buffer on the WASM side since those functions are obscured from the user, as far as I know -- we can keep it in mind as something to investigate in future releases. The C API takes in a pre-allocated buffer as one of its parameters.

File: js/common/lib/training-session.ts [hunk 7]
```diff
@@ -49,21 +50,31 @@ export interface TrainingSession {
...
    * @param buffer - buffer containing parameters
-   * @param trainableOnly - True if trainable parameters only to be modified, false otherwise.
+   * @param trainableOnly - True if trainable parameters only to be modified, false otherwise. Default value is true.
    */
-  loadParametersBuffer(array: Uint8Array, trainableOnly: boolean): Promise<void>;
+  loadParametersBuffer(array: Float32Array, trainableOnly: boolean): Promise<void>;
 
   /**
-   * Copies from the TrainingSession parameters to a buffer.
+   * Copies from the TrainingSession parameters to a contiguous buffer.
```
Comment (askhade) [inline js/common/lib/training-session.ts:71]:
```suggestion
   * Copies the model parameters to a contiguous buffer. Usually used in the context of Federated Learning
```

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 8]
```diff
@@ -313,6 +317,131 @@ export const runTrainStep = async(
...
+      throw new Error(NO_TRAIN_FUNCS_MSG);
+    }
+
+    const typedArrayConstructor = tensorTypeToTypedArrayConstructor(tensorTypeAsString);
+    const data = new typedArrayConstructor(parametersSize);
+    const output: TensorMetadata[] = [];
+    new Uint8Array(data.buffer, data.byteOffset, data.byteLength)
+        .set(wasm.HEAPU8.subarray(paramsOffset, paramsOffset + paramsByteLength));
+    output.push([tensorTypeAsString, dims, data, locationAsString]);
+    if (output.length > 1 || output.length < 1) {
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:385]: nit: output.length !== 1 ?

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 9]
```diff
@@ -313,6 +317,131 @@ export const runTrainStep = async(
...
+      throw new Error(`something unexpected happened in the getContiguousParameters function. Expected output length of
+     one, got ${output.length}`);
+    } else {
+      return output[0];
+    }
+  } finally {
+    if (tensor !== 0) {
+      wasm._OrtReleaseTensor(tensor);
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:393]: is this needed? ORT is not allocating the memory right? we allocate the memory in wasm right?

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 10]
```diff
@@ -313,6 +317,131 @@ export const runTrainStep = async(
...
+  const paramsByteLength = 4 * parametersSize;
+  const paramsOffset = wasm.stackAlloc(paramsByteLength);
+  wasm.HEAPU8.set(new Float32Array(parametersSize), paramsOffset);
+
+  const tensorOffset = wasm.stackAlloc(paramsOffset / 4);
+
+  // handles the dimensions-related createTensor parameters
+  const dims = [parametersSize];
+
+  const dimsOffset = wasm.stackAlloc(4);
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:359]: shouldn't this be 4 * parameterSize or paramsByteLength

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 11]
```diff
@@ -313,6 +317,131 @@ export const runTrainStep = async(
...
+      return wasm.HEAP32[sizeOffset / 4];
+    } else {
+      throw new Error(NO_TRAIN_FUNCS_MSG);
+    }
+  } finally {
+    wasm.stackRestore(stack);
+  }
+};
+
+export const getContiguousParameters =
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:339]: This can be part of a separate PR. I see there are lot of mtds which allocate data, it will be good to add some utils which can be called form everywhere, for example prepareInputOutputTensor also does a lot of this, not sure if you can simply call this mtd, but can do some refactoring to reduce duplication.

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 12]
```diff
@@ -313,6 +317,131 @@ export const runTrainStep = async(
...
+  const stack = wasm.stackSave();
+
+  const tensorTypeAsString = 'float32';
+  const locationAsString = 'cpu';
+
+  const parametersSize = getParametersSize(trainingSessionId, trainableOnly);
+  let tensor = 0;
+
+  const paramsByteLength = 4 * parametersSize;
+  const paramsOffset = wasm.stackAlloc(paramsByteLength);
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:351]: malloc?

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 13]
```diff
@@ -313,6 +317,130 @@ export const runTrainStep = async(
...
+  const tensorTypeAsString = 'float32';
+  const locationAsString = 'cpu';
+
+  const parametersSize = getParametersSize(trainingSessionId, trainableOnly);
+  let tensor = 0;
+
+  // allocates a Float32Array of the correct size on the WASM heap
+  const paramsByteLength = 4 * parametersSize;
+  const paramsOffset = wasm._malloc(paramsByteLength);
+  const arr = new Float32Array(parametersSize);
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:353]: I don't understand why we need this. Why are we allocating an array in JS?

Comment (carzh) [reply js/web/lib/wasm/wasm-training-core-impl.ts]: oops yes, you're right -- removed these two lines in the last commit.

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 14]
```diff
@@ -313,6 +317,128 @@ export const runTrainStep = async(
...
+  const dims = [parametersSize];
+
+  const dimsOffset = wasm.stackAlloc(4);
+  const dimsIndex = dimsOffset / 4;
+  wasm.HEAP32[dimsIndex] = parametersSize;
+
+  try {
+    // wraps allocated array in a tensor
+    tensor = wasm._OrtCreateTensor(
+        tensorDataTypeStringToEnum(tensorTypeAsString), paramsOffset, paramsByteLength, dimsOffset, dims.length,
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:364]:
```suggestion
        tensorDataTypeStringToEnum(tensorTypeAsString), paramsOffset, parametersSize, dimsOffset, dims.length,
```

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 15]
```diff
@@ -313,6 +317,128 @@ export const runTrainStep = async(
...
+    async(trainingSessionId: number, buffer: Float32Array, trainableOnly: boolean): Promise<void> => {
+  const wasm = getInstance();
+  const stack = wasm.stackSave();
+
+  const tensorTypeAsString = 'float32';
+  const locationAsString = 'cpu';
+
+  // allocates & copies JavaScript buffer to WASM heap
+  const bufferCount = buffer.length;
+  const bufferByteLength = bufferCount * 4;
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:410]: *4 not needed if you change the buffer to Uint8

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 16]
```diff
@@ -313,6 +317,128 @@ export const runTrainStep = async(
...
+
+  // allocates and handles moving dimensions information to WASM memory
+  const dimsOffset = wasm.stackAlloc(4);
+  wasm.HEAP32[dimsOffset / 4] = bufferCount;
+  const dimsLength = 1;
+  let tensor = 0;
+
+  try {
+    tensor = wasm._OrtCreateTensor(
+        tensorDataTypeStringToEnum(tensorTypeAsString), bufferOffset, bufferByteLength, dimsOffset, dimsLength,
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:422]: buffer.count

File: js/web/lib/wasm/wasm-training-core-impl.ts [hunk 17]
```diff
@@ -400,18 +398,18 @@ export const getContiguousParameters =
...
   const wasm = getInstance();
   const stack = wasm.stackSave();
 
   const tensorTypeAsString = 'float32';
   const locationAsString = 'cpu';
 
   // allocates & copies JavaScript buffer to WASM heap
-  const bufferCount = buffer.length;
-  const bufferByteLength = bufferCount * 4;
+  const bufferCount = getParametersSize(trainingSessionId, trainableOnly);
```
Comment (askhade) [inline js/web/lib/wasm/wasm-training-core-impl.ts:409]: nit: instead of calling this API, can you simply set this to buffer.length/4?

**Review by askhade — APPROVED** [review]
LGTM
